### PR TITLE
limit unix read/write to 2GB at a time

### PR DESCRIPTION
**Pull request by amitmurthy**
This is a libuv fix for https://github.com/JuliaLang/julia/issues/13578

~~I have not been able to test it as I have not yet got a handle how to incorporate this branch to the regular Julia build process. `make` in the Julia folder gives some git errors that I am unable to fix. This is after I edited libuv.version and also .git/config.  @vtjnash do you have a Julia libuv development workflow written down anywhere?~~

### CODE REVIEW

**Comment by tkelman** [comment]
what are the errors, exactly? you may need to `git fetch --unshallow` in `deps/srccache/libuv`, or try `make -C deps distclean-libuv` first

**Comment by amitmurthy** [comment]
Thanks. The `distclean` did the trick. It is already a complete repo hence the `unshallow` was not required.

**Comment by amitmurthy** [comment]
Julia test suite runs fine. Works for data send/recv > 2GB.

Addresses OSX part of https://github.com/JuliaLang/julia/issues/13578 

@vtjnash / @Keno , If OK please merge and bump deps/libuv.version.

This should be backported to Julia 0.4 too.

**Comment by Allardvm** [comment]
I like the clarify of this implementation, although I wonder if it isn't possible to prevent the extra memory allocation when writing large `iovec`s.

If we distinguish between the case when the sum of lengths causes an overflow and the case when a single iov causes an overflow, I believe we can prevent issues by limiting either the length of the buffer to write or the number of buffers to write. Something along following lines at [stream.c:797](https://github.com/JuliaLang/libuv/blob/julia-uv0.11.26/src/unix/stream.c#L797) should work (although the implementation isn't very elegant yet):

``` c
if (iovcnt == 1) {
  if (iov[0].iov_len > INT32_MAX) {
    n = write(uv__stream_fd(stream), iov[0].iov_base, INT32_MAX);
  } else {
    n = write(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len);
  }
} else {
  int newiov_cnt = 0;
  size_t total_bytes = 0;
  struct iovec* iov_ptr = iov;
  for (int i=0; i < iovcnt; i++) {
    if (total_bytes + iov_ptr->iov_len <= INT32_MAX) {
      newiov_cnt = newiov_cnt + 1;
      total_bytes = total_bytes + iov_ptr->iov_len;
    }
    iov_ptr++;
  }
  if (newiov_cnt == 0) {
    n = write(uv__stream_fd(stream), iov[0].iov_base, INT32_MAX);
  } else {
    n = writev(uv__stream_fd(stream), iov, newiov_cnt);
  }
}
```

The existing `uv__write` code should detect that `n` does not cover all the bytes that should have been written and leave whatever still needs to be written in the queue.

Note that this doesn't close https://github.com/JuliaLang/julia/issues/13578, since the initial minimal working example will still produce a BSOD on a Windows machine.

File: src/unix/stream.c [hunk 1]
```diff
@@ -794,10 +825,21 @@ static void uv__write(uv_stream_t* stream) {
     while (n == -1 && errno == EINTR);
   } else {
     do {
-      if (iovcnt == 1) {
-        n = write(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len);
+      /* On some platforms, notably OSX, the sum of iov_len must not
+       * overflow a 32-bit integer.
+       */
+
+      int newiovcnt = iovcnt;
```
Comment (ScottPJones) [inline src/unix/stream.c]: Shouldn't this be limited to the platforms where it pwritev/writev are actually "broken" (i.e. Mac OS X)?

**Comment by vtjnash** [comment]
I think it should be valid to avoid the extra allocation by temporarily directly modifying iovec:

``` julia
int64_t sum = 0;
for (i in iovec)
    if (sum + length > INT32_MAX)
        oldsize = i.size
        i.size = INT32_MAX - sum
        break
    sum += length(i)
end
writev(iovec, i)
if i < length(iovec)
    i.size = oldsize
```

i doubt the kernel will be able to atomically write 2 GB of data without blocking, so it shouldn't be too big of a concern that this PR will cause libuv to re-enter the event loop thinking that the write was incomplete (e.g. that the kernel would issue EAGAIN if another write was attempted).

as an optimization, `write_queue_size` can be checked at the top of the loop to decide if it can be short-circuited in the common-case of writing < INT32_MAX bytes of data.

the loop should probably only be active for 32-bit machines (`sizeof(size_t) < 8`) and Darwin (`ifdef __APPLE__`):

``` c
if (
#ifdef __APPLE__
1
#else
sizeof(size_t) < 8
#endif
)
```

**Comment by amitmurthy** [comment]
I prefer @Allardvm 's suggestion. Directly modifying iov array would necessitate setting subsequent iov_len's to 0, which means we will have to track multiple oldsizes. And probably dealing with platforms that do not like zero values for iov_len's.

Checking `write_queue_size` is a good idea, though I would leave out the `#ifdefs`, Code subsequent to the `write` tracks which buffers have been written and by how much. Practically speaking I have seen the `write` calls actually returning after every 64K of data written out on TCP sockets. On larger machines (or file io), the overhead of a syscall for every 2GB write will still be very, very small.

Will update the PR. 

Since I do not have access to a Windows machine will prefer that to be fixed separately.

**Comment by vtjnash** [comment]
> iov_len's to 0, which means we will have to track multiple oldsizes

in the code above, i was intending to imply that the number of iovec's would also be adjusted (libuv already has code that does this), so only one iovec entry would get modified for a given call to writev

**Comment by amitmurthy** [comment]
Right. My bad. Good suggestion.

**Comment by vtjnash** [comment]
when you rebase, it's also worth taking a look at the libuv style guide (https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md#commit). they are pretty strict about things like two newlines between functions and the formatting of the commit messages.

**Comment by ScottPJones** [comment]
An important thing is that this should not be done for file operations - from experience, I know that the file operations are atomic wrt other writers, even for well over 2GB (tested this for when different processes were all extending the same database file).  Internally, the OS locks changes to the file metadata for the duration.
Also, there is a limit on the number of iovec's permitted, usually 1K.  If you are going to be increasing the number of iovecs, that system limit needs to be checked.

**Comment by amitmurthy** [comment]
Done and tested.

**Comment by amitmurthy** [comment]
Merging this in a day.

I assume that 
- merging into `julia-uv0.11.26` is the correct thing to do
- will get activated in the next libuv update cycle.

**Comment by Allardvm** [comment]
Lastly, we could try to base the `iov_len` limits on the POSIX specification and use `SSIZE_MAX` instead of `INT32_MAX` to follow up on @vtjnash's suggestion that this isn't needed everywhere . We would have to check whether this works on Darwin 64bit, since I haven't found a definite source on how `SSIZE_MAX` is defined there. We can use `INT32_MAX` as a safe fallback by including something along the following lines:

``` C
#ifndef SSIZE_MAX
#define SSIZE_MAX INT32_MAX
#endif
```

**Comment by amitmurthy** [comment]
On Darwin 64-bit `SSIZE_MAX` is 2^63-1.

```
#include <stdio.h>
#include <limits.h>

int main()
{
  printf("%ld\n", SSIZE_MAX);
  return 0;
}
```

It is a specific documented 2GB limitation in the implementation of read/write on OSX (and maybe the BSDs), though POSIX states otherwise.

**Comment by Allardvm** [comment]
Thanks for looking it up! If `SSIZE_MAX` doesn't work, fixing it to `INT32_MAX` might indeed be the best we can do. Golang only implements the fix for FreeBSD and Darwin, but I agree that it is unlikely to make a relevant impact on the performance on other platforms if we do it everywhere.

For completeness sake, note that Golang eventually implemented a second fix over the initial one to limit reads and writes to 1GB (`1 << 30`) instead of 2GB - 1 to ensure subsequent reads and writes remain aligned. I am not nearly enough of an I/O expert to say anything about whether the performance hit that results from this is relevant. Considering that the Golang devs thought it was significant enough to warrant fixing, I propose that we just follow their example. I'm happy with either implementation though.

**Comment by amitmurthy** [comment]
Updated.

File: src/unix/stream.c [hunk 2]
```diff
@@ -794,11 +795,36 @@ static void uv__write(uv_stream_t* stream) {
...
+
+      if (stream->write_queue_size > INT32_MAX) {
+        size_t total_bytes = 0;
+        int new_iov_cnt=0;
+
+        for (int i=0; i < iovcnt; i++) {
+          new_iov_cnt++;
+          old_iov_len = iov[i].iov_len;
+          if ((total_bytes + iov[i].iov_len) >= INT32_MAX) {
+            old_iov_len = iov[i].iov_len;
```
Comment (Allardvm) [inline src/unix/stream.c]: This line (811) seems redundant, considering it is updated right before the `if` statement. Everything else looks great!

Comment (amitmurthy) [inline src/unix/stream.c]: Updated. Thanks.

**Comment by ScottPJones** [comment]
Is the limitation there for`write`, or just `writev`, on OS X?  I only found documentation of a limitation for the `writev` call.

**Comment by amitmurthy** [comment]
Yes the doc is only for writev and readv. But calls to both write and read were also failing with EINVAL for lengths > 2GB.

**Comment by ScottPJones** [comment]
Direct calls to `write` on files in OS X, and not through the `libuv` `uv_fs_write` call?  That's not good!

**Comment by amitmurthy** [comment]
I just tested for TCP Sockets. It is the `write` call in the code changes in this PR (`src/unix/stream.c`). Don't know if the same code path is used for file writes.

**Comment by Allardvm** [comment]
Writing to a file with an `IOStream` follows a different path, but ultimately also calls `read` and `write` (not the vector versions). Despite the lack of documentation, large file reads and writes have previously lead to errors on Darwin (see https://github.com/JuliaLang/julia/pull/6483 and https://github.com/JuliaLang/julia/commit/0c1100412f509a6e1e75750a0fab235009ea47a3) and on Windows (see https://github.com/JuliaLang/julia/pull/8689).

**Pull request merged**